Mathlib/Algebra/Lie/LieTheorem.lean

```diff
@@ -0,0 +1,240 @@
+/-
+Copyright (c) 2024 Lucas Whitfield. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lucas Whitfield, Johan Commelin
+-/
+import Mathlib.Algebra.Lie.Weights.Basic
+import Mathlib.RingTheory.Finiteness.Nilpotent
+
+/-!
+# Lie's theorem for Solvable Lie algebras.
+
+Lie's theorem asserts that Lie modules of solvable Lie algebras over fields of characteristic 0
+have a common eigenvector for the action of all elements of the Lie algebra.
+This result is named `LieModule.exists_forall_lie_eq_smul_of_isSolvable`.
+-/
+
+namespace LieModule
+
+section
+
+/-
+The following variables generalize the setting where:
+- `R` is a principal ideal domain of characteristic zero,
+- `L` is a Lie algebra over `R`,
+- `V` is a Lie algebra module over `L`
+- `A` is a Lie ideal of `L`.
+Besides generalizing, it also make the proof of `lie_stable` syntactically smoother.
+-/
+variable {R L A V : Type*} [CommRing R]
+variable [IsPrincipalIdealRing R] [IsDomain R] [CharZero R]
+variable [LieRing L] [LieAlgebra R L]
+variable [LieRing A] [LieAlgebra R A]
+variable [LieRingModule L A] [LieRingModule A L]
+variable [AddCommGroup V] [Module R V] [Module.Free R V] [IsNoetherian R V]
+variable [LieRingModule L V] [LieModule R L V]
+variable [LieRingModule A V] [LieModule R A V]
+variable [IsLieTower L A V] [IsLieTower A L V]
+
+variable (χ : A → R)
+
+open Module (finrank)
+open LieModule
+
+local notation "π" => LieModule.toEnd R _ V
+
+private abbrev T (w : A) : Module.End R V := (π w) - χ w • 1
+
```

Comment:
It seems like we shouldn't need this.
```suggestion
```
Can't we just add:
```lean
  set T : A → Module.End R V := fun w ↦ (π w) - χ w • 1 with hT
```
just before we start talking about `T`, replace a bunch of `T χ w` with `T w` and use `hT` to do some extra simplification where necessary?

Reply:
No, we are using `T` with two different `V`s in the the proof. Mostly just `V`, but once it is `U`.

Reply:
I see, thanks.